### ISSUE TITLE
Assembly Version Info does not propagate into output artifacts for release builds

### ISSUE THREAD

**Issue by jonorossi** [castleproject/Core#206]
Running the TeamCity "Pack" build configuration as a release build with a version and version suffix won't cause dependencies to be rebuilt with that configuration.

* All assemblies are still 0.0.0.0
* The copyright year isn't populated in assemblies

This must be fixed before v4.0 final.

**Comment by jonorossi** [comment]
@blairconrad @jeremymeng not sure how we missed this when we made the changes to versioning and packaging. I'm not sure how to fix this, TeamCity does have an option to on the "run custom" build dialog for "rebuild all snapshot dependencies transitively", however I don't think it'll pass parameters into the dependencies.

Thoughts?

**Comment by blairconrad** [comment]
This sounds bad. @jonorossi, I forget the setup for the build. Lemme run off and refresh myself.

**Comment by blairconrad** [comment]
@jonorossi, perhaps I'm naive, but would running Set-Version before each build on the TeamCity server do the trick?

**Comment by blairconrad** [comment]
I think I _am_ naive, as `Set-VersionInfo` appears to be the first step in all the builds.

Except for 4.5, for which `Set-VersionInfo` is the _second_ step.

Oh, and Mono. Which I think is only for tests. Although I've never fully understood this.

**Comment by blairconrad** [comment]
I think that's the problem, @jonorossi. Check out this from the 4.0-beta001 build:

![image](https://cloud.githubusercontent.com/assets/3275797/16912768/26939e82-4cb4-11e6-9c5a-499ed72989cc.png)

I'm going to use my initiative and reorder the build steps for 4.5. You can thank me or tell me off after! :-)

**Comment by blairconrad** [comment]
Okay. So if the `Settings.proj` is edited, any Pack will generate a nupkg with good values.

I suppose, though, if we change the `$version_suffix` or `$release_build` and then just run Pack, it won't rebuild all the libs. (Unless TeamCity will rebuild when we change the environment vars - this is not something I know offhand.)

If this is a concern, we could edit the Pack configuration so it will _always_ rebuild the 4 flavours of the library and not just use a version that it thinks is good. It'll slow things down from time to time, but would be safer.

**Comment by blairconrad** [comment]
In case you're curious, after rearranging the steps, then building 4.5 and then running Pack, I get this from the nupkg in the artifacts:

![image](https://cloud.githubusercontent.com/assets/3275797/16913479/a7aee1b2-4cb8-11e6-994a-83e122ba26c0.png)

**Comment by jonorossi** [comment]
> Except for 4.5, for which Set-VersionInfo is the second step.

Oops, yep it was meant to be first. I manually added that step to each configuration before merging the netcore branch, thanks for fixing that.

Copyright is sorted, but the version is still a problem. e.g. this is what the build will get because each build configuration doesn't rebuild.

```
[assembly: AssemblyInformationalVersion("4.0.0-ci02340")]
```

I'll see if we can get parameters to pass through the build chain.

**Comment by jonorossi** [comment]
@blairconrad sorry for the errors in my issue description, I logged this issue at 3.30am.

**Comment by blairconrad** [comment]
> I'll see if we can get parameters to pass through the build chain.

I don't quite follow. Right now, the `Pack` won't trigger a rebuild if the source hasn't changed, so we'll keep the `[assembly: AssemblyInformationalVersion("4.0.0-ci02340")]` attribute when the nupkg is built (and technically the year could be wrong too, if we build the lib in 2016 and the nupkg in 2017).

If we toggle "Do not run new build if there is a suitable one", then the DLLs will be regenerated. (In this case, I'd consider removing the "build on VCS" trigger from the 4 DLL configurations, though.)

![image](https://cloud.githubusercontent.com/assets/3275797/16913821/7353d6a4-4cbb-11e6-8dfc-4a39b0155be8.png)

And if a build is forced, it should get the same values from the Settings.proj and the environment vars, no?

**Comment by jonorossi** [comment]
The Pack configuration has other properties defined which don't seem needed, only `release_build` and `version_suffix` are used.

It [looks like TeamCity can do what we want](https://confluence.jetbrains.com/display/TCD9/Predefined+Build+Parameters#PredefinedBuildParameters-OverridingDependenciesProperties), you define `reverse.dep.*.<property name>` or `reverse.dep.<btID>.<property name>` to override dependency properties which will force them to be rebuilt with the overridden parameters.

I'll give it a go.

**Comment by jonorossi** [comment]
> And if a build is forced, it should get the same values from the Settings.proj and the environment vars, no?

Nope, see my previous comment about why it won't get the environment's properties because they are only set further along the build chain at the moment.

**Comment by blairconrad** [comment]
> > And if a build is forced, it should get the same values from the Settings.proj and the environment vars, no?
> Nope, see my previous comment about why it won't get the environment's properties because they are only set further along the build chain at the moment.

Ah. No. I still don't get it. Perhaps I'm slow. If `Pack` always forces a new build of (say) 4.5, then will 4.5 not be built with the current… 

I just got it! There's no `version_suffix` or `release_build` property on the libraries. You're going to use the fancy [Overriding Dependencies Properties](https://confluence.jetbrains.com/display/TCD9/Predefined+Build+Parameters#PredefinedBuildParameters-OverridingDependenciesProperties) to supply those settings where they didn't already exist. And I see from the docs that this will force a build when needed.

Yes. That sounds very promising. Sorry I didn't realize back in the day that this would be a problem.

**Comment by jonorossi** [comment]
@blairconrad exactly, well that's the plan. Now to work out how to mix configuration parameters and environment variables without making it a confusing mess.

**Comment by blairconrad** [comment]
Not suggesting a change, but as an aside … I've always favoured just slamming the versions (including pre-release markers, if any) in the code. That would avoid this problem. On the other hand, sometimes people find the friction of committing `-beta` change to be too much. Which I can also understand.

**Comment by jonorossi** [comment]
@blairconrad I'd actually prefer to actually pull the version out of code completely. Maybe I'm a purist but it is called a "version control" system for a reason, the version should be derived from the Git tag, however I'm not going down that path just yet as the tooling is very messy here, so much history.

**Comment by jonorossi** [comment]
I've found an even easier way to do these parameters, meet [Project Level Build Parameters](https://confluence.jetbrains.com/display/TCD9/Project+and+Agent+Level+Build+Parameters) which each build configuration will inherit. Running a test build now.

**Comment by blairconrad** [comment]
That sounds great! Although it's not clear to me that a changed project level parameter will force the rebuild of the libs when you build Pack. I guess you'll find out!

**Comment by blairconrad** [comment]
About versions, and this is not really the place for this conversation, getting the version from the Git tag is just a kind of getting it from the code. They both depend on permanent changes being shoved into the VCS somehow. For me, I've always preferred the explicitness of these things living in files (instead of Git tags or SVN properties or similar) because they are more visible, the people I've worked with have traditionally been unsophisticated with regard to the VCS powers, and sometimes we've switched VCSes at work. Then there's always extra rework of the build system to get things to work.

Not trying to change your mind. Just chatting. I'll leave you alone.

**Comment by jonorossi** [comment]
@blairconrad it does, TeamCity keeps track of what build parameters are used for each build, if a snapshot dependency doesn't have the same parameters it'll have to rebuild it. I've also updated the descriptions and UI a bit for the custom parameters so the custom run build dialog is a bit easier to follow.

**Comment by blairconrad** [comment]
Very fancy. I'm glad it's working out. Sorry about the deficiency.

**Comment by jonorossi** [comment]
> About versions, and this is not really the place for this conversation, getting the version from the Git tag is just a kind of getting it from the code. They both depend on permanent changes being shoved into the VCS somehow. For me, I've always preferred the explicitness of these things living in files (instead of Git tags or SVN properties or similar) because they are more visible, the people I've worked with have traditionally been unsophisticated with regard to the VCS powers, and sometimes we've switched VCSes at work. Then there's always extra rework of the build system to get things to work.
> 
> Not trying to change your mind. Just chatting. I'll leave you alone.

No problem at all. My view on this is that every commit causes the version to change, i.e. the current HEAD of master isn't 4.0.0-beta001 it is already further along after that version so having a file in the repo that tells you that is the version of the code is sort of wrong, just as switching your version file to 4.0.0 way before making a 4.0.0 release is. Also from the purist perspective, if you are doing RCs then the Git tag for the final release should actually be the exact same Git ref as the RC you chose.

As you said this isn't the forum, however I'm happy to chat on these topics.

**Comment by blairconrad** [comment]
> Also from the purist perspective, if you are doing RCs then the Git tag for the final release should actually be the exact same Git ref as the RC you chose.

It's true. Then again, if there are too many external factors that touch the build, I lose confidence in the Git refs!

So many competing factors.

**Comment by jonorossi** [comment]
Fixed. With a bit of playing around I worked out that I could set the environment variables to resolve to `reverse.dep.*.<property name>` new configuration parameters and make the environment variables hidden fields and the configuration parameters the visible ones. It is a bit strange underneath but it all comes out right in the builds. These reverse.dep build parameters are a bit strange at times as they resolve when the build is added to the queue not when that build starts in the chain as obviously the values have to be passed up the chain.
